### PR TITLE
Add multinomial sampling to sequence generation

### LINKED ISSUE #35: Support more text generation strategies
Currently `Bumblebee.Text.Generation.generate/5` supports only a basic greedy strategy for selecting the next token. We should add options for more sophisticated strategies, in particular beam search and sampling.

[Here](https://huggingface.co/blog/how-to-generate) is a good reference explaining various options. And [contrastive search](https://huggingface.co/blog/introducing-csearch), a more recent development.

### PR DESCRIPTION
**Pull request by seanmor5**
Adds multinomial sampling strategy and top-k/top-p options for adjusting the distribution.

### CODE REVIEW

File: lib/bumblebee/text/generation.ex [hunk 1]
```diff
@@ -121,6 +121,17 @@ defmodule Bumblebee.Text.Generation do
...
+    * `:num_beams` - the number of beams to use in a beam search. If set
+      to 1, beam search will not be used. Defaults to `1`
+
+    * `:early_stopping` - whether to stop the beam search when at least
+      `:num_beams` sentences are finished per batch or not. Defaults to
+      `false`
+
+    * `:sample` - whether or not to use random sampling. Defaults to `false`
+
+    * `:prng_key` - random key to use when sampling. Defaults to `nil`
```
Comment (jonatanklosko) [inline lib/bumblebee/text/generation.ex:133]: Since this function is pretty higher-level, what about `:seed` instead?

File: lib/bumblebee/text/generation.ex [hunk 2]
```diff
@@ -507,6 +555,134 @@ defmodule Bumblebee.Text.Generation do
...
+        sequences: sequences,
+        length: length,
+        input_length: input_length
+      })
+
+    # TODO: logits warper
+
+    vocab = Nx.iota(Nx.shape(logits), axis: -1)
+    probabilities = Axon.Activations.softmax(logits)
+    next_token = Nx.Random.choice(key, vocab, probabilities)
```
Comment (josevalim) [inline lib/bumblebee/text/generation.ex:672]:
```suggestion
    next_token = Nx.Random.choice(key, vocab, probabilities, [])
```

File: lib/bumblebee/text/generation.ex [hunk 3]
```diff
@@ -507,6 +555,134 @@ defmodule Bumblebee.Text.Generation do
...
+          )
+
+        {sequences, length, finished?, inputs, params, prng_key}
+      end
+
+    sequences
+  end
+
+  defnp sample_condition(finished?, length, max_length) do
+    Nx.logical_not(Nx.logical_or(Nx.all(finished?), Nx.equal(length, max_length)))
```
Comment (josevalim) [inline lib/bumblebee/text/generation.ex:635]:
```suggestion
    not(Nx.all(finished?) or length == max_length)
```

?

File: lib/bumblebee/text/generation.ex (hunk 1 above)
Comment (seanmor5) [reply lib/bumblebee/text/generation.ex:133]: Agreed, though one thing I was thinking was if we should pass this in the serving as well and return the updated key. They do accept `key` in `generate` in HF, but they do not return an updated key, so maybe just allowing a key to the serving is enough and it is assumed you will split it outside the serving

File: lib/bumblebee/text/generation.ex [hunk 4]
```diff
@@ -507,6 +555,134 @@ defmodule Bumblebee.Text.Generation do
...
+        sequences: sequences,
+        length: length,
+        input_length: input_length
+      })
+
+    # TODO: logits warper
+
+    vocab = Nx.iota(Nx.shape(logits), axis: -1)
+    probabilities = Axon.Activations.softmax(logits)
+    next_token = Nx.Random.choice(key, vocab, probabilities, [])
```
Comment (josevalim) [inline lib/bumblebee/text/generation.ex:672]: Actually, you need to choose the number of samples.

File: lib/bumblebee/text/generation.ex (hunk 1 above)
Comment (jonatanklosko) [reply lib/bumblebee/text/generation.ex:133]: Servings are so high-level that it seems to me the seed is for reproducibility if anything else (or to explore different result for the same prompt), and so a number is more transparent than a tensor key. Or do you think there's a scenario where you would actually use a split key?

Comment (seanmor5) [reply lib/bumblebee/text/generation.ex]: Sorry, what I mean is currently you'd need to create a new serving with a new seed to get different inputs. Maybe this is desirable though, but I'm not sure as if you want something non-deterministic you may want the option to pass a key to the serving

File: lib/bumblebee/text/generation.ex [hunk 5]
```diff
@@ -507,6 +540,147 @@ defmodule Bumblebee.Text.Generation do
...
+        ) do
+    pad_token_id = opts[:pad_token_id]
+    eos_token_id = opts[:eos_token_id]
+
+    key = Nx.Random.split(prng_key)
+    {key, key_next} = {key[1], key[0]}
+
+    model_outputs = predict_fun.(params, inputs)
+
+    logits = model_outputs.logits[[0..-1//1, -1]]
```
Comment (msluszniak) [inline lib/bumblebee/text/generation.ex:644]:
```suggestion
    logits = model_outputs.logits[[.., -1]]
```

File: lib/bumblebee/text/generation.ex [hunk 6]
```diff
@@ -507,6 +540,147 @@ defmodule Bumblebee.Text.Generation do
...
+
+    {sequences, length + 1, next_finished?, inputs, key_next}
+  end
+
+  deftransformp batched_choice(key, vocab, probabilities) do
+    {batch_size, _} = Nx.shape(vocab)
+
+    tokens =
+      for i <- 0..(batch_size - 1) do
+        candidate_tokens = vocab[[i, 0..-1//1]]
```
Comment (msluszniak) [inline lib/bumblebee/text/generation.ex:673]:
```suggestion
        candidate_tokens = vocab[[i, ..]]
```

File: lib/bumblebee/text/generation.ex [hunk 7]
```diff
@@ -507,6 +540,147 @@ defmodule Bumblebee.Text.Generation do
...
+        candidate_probs = probabilities[[i, 0..-1//1]]
```
Comment (msluszniak) [inline lib/bumblebee/text/generation.ex:674]:
```suggestion
        candidate_probs = probabilities[[i, ..]]
```

**Comment by jonatanklosko** [comment]
@seanmor5 I finished sampling, we can do beam search separately, especially that we have contrastive search too :)

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
> we can do beam search separately

If you think about it, every search we do is a _beam_ search.
